core/src/test/scala/unit/kafka/coordinator/transaction/ProducerIdManagerTest.scala

```diff
@@ -38,22 +38,49 @@ import org.mockito.Mockito.{mock, when}
 import java.util.concurrent.{CountDownLatch, Executors, TimeUnit}
 import java.util.concurrent.atomic.AtomicBoolean
 import scala.collection.mutable
-import scala.util.{Failure, Success}
+import scala.util.{Failure, Success, Try}
 
 class ProducerIdManagerTest {
 
   var brokerToController: NodeToControllerChannelManager = mock(classOf[NodeToControllerChannelManager])
   val zkClient: KafkaZkClient = mock(classOf[KafkaZkClient])
 
+  case class ErrorCount(error: Errors, var repeat: Int)
+
+  object ErrorCount {
+    def once(error: Errors): ErrorCount = {
+      ErrorCount(error, 1)
+    }
+  }
+
+  class ErrorQueue(initialErrorCounts: ErrorCount*) {
+    private val queue: mutable.Queue[ErrorCount] = mutable.Queue.empty ++ initialErrorCounts
+
+    def takeError(): Errors = queue.synchronized {
+      if (queue.isEmpty) {
+        return Errors.NONE
+      }
+      while (queue.nonEmpty && queue.head.repeat == 0) {
+        queue.dequeue()
+      }
+      if (queue.isEmpty) {
+        return Errors.NONE
+      }
+      if (queue.head.repeat > 0) {
+        queue.head.repeat -= 1
+      }
+      queue.head.error
```

Comment:
Should the error be immediately dequeued once it's taken?

Reply:
i just removed the maybe request next block and it required peeking into it, but since it's gone, no need to have the counters.